### ISSUE TITLE
Gatsby Image plugin - Constrained layout - Incorrect dimensions

### ISSUE THREAD

**Issue by ramiroazar** [gatsbyjs/gatsby#32799]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report directly related to Gatsby. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Description

When using either `StaticImage` or `GatsbyImage` components with the default `layout` prop value `constrained`, the image loads with incorrect dimensions when navigated to from another page using `Link` component, which causes layout shifts.

### Reproduction Link

https://github.com/ramiroazar/gatsby-starter-default/tree/issue/gatsby-plugin-image/constrained

### Steps to Reproduce

1. Load http://localhost:8000/
2. Check correct dimensions logged in console
3. Load http://localhost:8000/page-2/
4. Navigate to http://localhost:8000/
5. Check incorrect dimensions logged in console

### Expected Result

The image should load with the same dimensions when the page is loaded directly and navigated to using the `Link` component.

In the case of this example that should be `{ width: 920.875, height: 326.09375 }`.

### Actual Result

The image loads with incorrect dimensions when the page is navigated to using the`Link` component.

In the case of this example it's logging `{ width: 0, height: 26.09375 }` when navigated to from another page.

Note that this only happens if the page hasn't been previously loaded and is navigated to using the `Link` component for the first time.

### Environment

```shell
System:
  OS: macOS 11.4
  CPU: (4) x64 Intel(R) Core(TM) i5-7500 CPU @ 3.40GHz
  Shell: 3.2.57 - /bin/bash
Binaries:
  Node: 14.16.1 - ~/.nvm/versions/node/v14.16.1/bin/node
  npm: 6.14.12 - ~/.nvm/versions/node/v14.16.1/bin/npm
Languages:
  Python: 2.7.16 - /usr/bin/python
Browsers:
  Chrome: 92.0.4515.131
  Firefox: 90.0.2
  Safari: 14.1.1
npmPackages:
  gatsby: ^3.11.1 => 3.11.1
  gatsby-plugin-gatsby-cloud: ^2.11.0 => 2.11.0
  gatsby-plugin-image: ^1.11.0 => 1.11.0
  gatsby-plugin-manifest: ^3.11.0 => 3.11.0
  gatsby-plugin-offline: ^4.11.0 => 4.11.0
  gatsby-plugin-react-helmet: ^4.11.0 => 4.11.0
  gatsby-plugin-sharp: ^3.11.0 => 3.11.0
  gatsby-source-filesystem: ^3.11.0 => 3.11.0
  gatsby-transformer-sharp: ^3.11.0 => 3.11.0
```

### Config Flags

DEV_SSR: true

**Comment by ramiroazar** [comment]
Yes, I can confirm this happens with both `gatsby develop` and `gatsby build` using in the reproduction link above.

**Comment by EyaChebbi** [comment]
Hello is anyone working on this issue ? Is it already fixed ?
